commerce-service/src/main/java/com/liferay/commerce/internal/upgrade/v4_1_0/CommerceAddressUpgradeProcess.java

```diff
@@ -14,21 +14,141 @@
 
 package com.liferay.commerce.internal.upgrade.v4_1_0;
 
+import com.liferay.commerce.account.model.CommerceAccount;
+import com.liferay.commerce.constants.CommerceAddressConstants;
 import com.liferay.commerce.internal.upgrade.base.BaseCommerceServiceUpgradeProcess;
 import com.liferay.commerce.model.impl.CommerceAddressImpl;
 import com.liferay.commerce.model.impl.CommerceAddressModelImpl;
+import com.liferay.portal.kernel.service.ClassNameLocalService;
+import com.liferay.portal.kernel.upgrade.UpgradeException;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
 
 /**
  * @author Alec Sloan
  */
 public class CommerceAddressUpgradeProcess
 	extends BaseCommerceServiceUpgradeProcess {
 
+	public CommerceAddressUpgradeProcess(
+		ClassNameLocalService classNameLocalService) {
+
+		_classNameLocalService = classNameLocalService;
+	}
+
 	@Override
 	protected void doUpgrade() throws Exception {
+		if (!hasColumn("CommerceAccount", "defaultBillingAddressId") ||
+			!hasColumn("CommerceAccount", "defaultShippingAddressId")) {
+
+			throw new UpgradeException(
+				"Upgrade fails as CommerceAccount does not have both " +
+					"columns: 'defaultBillingAddressId' and" +
+						"'defaultShippingAddressId'");
+		}
+
+		addColumn(
+			CommerceAddressImpl.class, CommerceAddressImpl.TABLE_NAME, "type",
+			"INTEGER");
+
+		PreparedStatement ps = null;
+
 		if (hasColumn(CommerceAddressModelImpl.TABLE_NAME, "groupId")) {
 			dropColumn(CommerceAddressImpl.TABLE_NAME, "groupId");
 		}
+
+		if (hasColumn(CommerceAddressModelImpl.TABLE_NAME, "defaultBilling")) {
+			ps = connection.prepareStatement(
+				"update CommerceAccount set defaultBillingAddressId = ? " +
+					"where commerceAccountId = ?");
+
+			updateCommerceAccountAndSetType(
+				ps, getCommerceAddressResultSet("defaultBilling"));
+
+			dropColumn(CommerceAddressImpl.TABLE_NAME, "defaultBilling");
```

Comment:
since we are keeping those columns to avoid version change, we should not drop it